File: sudoku.js
Array.range = (start, end) => Array.from({length: (end - start)}, (v, k) => k + start);

function isNumberKey(evt) {
    var charCode = (evt.which) ? evt.which : event.keyCode
    if (charCode > 31 && (charCode < 49 || charCode > 57))
        return false;    
    return true;
}

function locate_cell(board) {
    for (var j = 0; j < board.length; j++) {
        for (var i = 0; i < board[0].length; i++) {
            if (board[j][i] == 0) {
                return [j, i];
            }
        }
    }
    return null;
}

function check_guess(board, position, guess) {
    var y = position[0],
        x = position[1];
    
    for (var i = 0; i < board[0].length; i++) {
        if (board[y][i] == guess && x != i){
            return false;
        }
    }
    
    for (var j = 0; j < board.length; j++) {
        if (board[j][x] == guess && y != j){
            return false;
        }
    }
    
    var sect_x = Math.floor(x/3)
    var sect_y = Math.floor(y/3)
    
    sect_y_range = Array.range(sect_y*3, sect_y*3 + 3)
    sect_x_range = Array.range(sect_x*3, sect_x*3 + 3)
    
    for (j in sect_y_range) {
        for (i in sect_x_range) {
            if (board[sect_y_range[j]][sect_x_range[i]] == guess && [sect_x_range[i], sect_y_range[j]] != position) {
                return false;
            }
        }
    }
    
    return true;
}

function find_solution(board) {
    var position = locate_cell(board)
    
    if (position == null) {
        return true;
    } else {
        var y = position[0],
            x = position[1];
    }
    
    values = Array.range(1,10)
    attempts = 0;
    for (i in values) {
        if (check_guess(board, [y,x], values[i]) == true) {
            board[y][x] = values[i];
            
            if (find_solution(board) == true) {
                return true;
            }
            
            board[y][x] = 0;
        }
        attempts += 1;
        if (attempts > 50) {
            return false
        }
    }  
}

document.getElementById('solve').addEventListener('click', function(){
    var errmsg = document.getElementById("errmsg");
    errmsg.classList.add("hidden");
    var valueArr = [];
    var totalArr = [];
    var enteredVals =[]
    count = 0;
    rowcount = 0;
    document.querySelectorAll('.aa').forEach(function(el){
        if (el.value == "") {
            valueArr.push(0);
        } else {
            valueArr.push(parseInt(el.value));
            enteredVals.push([rowcount, count])
        }
        count += 1;
        if (count == 9) {
            totalArr.push(valueArr);
            rowcount += 1
            valueArr = [];
            count = 0;
        }
    });
    
    countx = 0;
    county = 0;
    countvals = 0;
    console.log(JSON.parse(JSON.stringify(enteredVals)));
    if (find_solution(totalArr) == true){
        document.querySelectorAll('.bb').forEach(function(el){
            if (countvals < enteredVals.length && county == enteredVals[countvals][0] && countx == enteredVals[countvals][1]) {
                el.innerHTML = totalArr[county][countx];
                el.classList.add("enteredVal")
                countvals += 1;
            } else {
                el.innerHTML = totalArr[county][countx];
            }
            
            countx += 1;
            if (countx == 9) {
                countx = 0;
                county += 1;
            }
        });
        var solution = document.getElementById("solution");
        solution.classList.remove("hidden");
        var inputbox = document.getElementById("inputbox");
        inputbox.classList.add("hidden");
    } else {
        var errmsg = document.getElementById("errmsg");
        errmsg.classList.remove("hidden");
        errmsg.classList.add("fade-in")
    };
});

document.getElementById('reset').addEventListener('click', function(){
    location.reload();
});

document.getElementById('clear').addEventListener('click', function(){
    location.reload();
});

File: sudoku.py
board1 = [
    [5,0,0,0,0,0,0,0,2],
    [0,8,0,6,0,0,0,0,5],
    [0,0,0,5,3,4,0,7,0],
    [0,7,0,4,0,0,9,0,0],
    [0,0,8,0,5,0,6,0,0],
    [0,0,1,0,0,3,0,5,0],
    [8,1,3,9,2,0,0,0,0],
    [2,0,0,3,0,1,0,8,0],
    [7,0,0,0,4,0,0,0,3]
]

def display_board(board):
    for j in range(len(board)):
        if j % 3 == 0 and j != 0:
            print("- - - - - - - - - - -")
            
        for i in range(len(board[0])):
            if i % 3 == 0 and i != 0:
                print("| ", end="")
            if i == len(board)-1:
                print(board[j][i])
            else:
                print(str(board[j][i]), end=" ")
    return
                
def locate_cell(board):
    for j in range(len(board)):
        for i in range(len(board[0])):
            if board[j][i] == 0:
                return (j, i)
    return None

def check_guess(board, position, guess):
    y,x = position
    
    for i in range(len(board[0])):
        if board[y][i] == guess and x != i:
            return False
    
    for j in range(len(board)):
        if board[j][x] == guess and y != j:
            return False
        
    sect_x = x // 3
    sect_y = y // 3
    
    for j in range(sect_y*3, sect_y*3 + 3):
        for i in range(sect_x*3, sect_x*3 + 3):
            if board[j][i] == guess and (i,j) != position:
                return False
        
    return True

def find_solution(board):
    find = locate_cell(board)
    if not find:
        return True
    else:
        y,x = find
        
    for i in range(1,10):
        if check_guess(board, (y,x), i):
            board[y][x] = i
            
            if find_solution(board):
                return True
        
            board[y][x] = 0
    return